### PR TITLE
Retry cleaned up fix for issue #541

### PR DESCRIPTION
**Pull request by srampal**

### CODE REVIEW

File: netctl/netctl.go [hunk 1]
```diff
@@ -65,7 +65,7 @@ func inspectPolicy(ctx *cli.Context) {
 	tenant := ctx.String("tenant")
 	policy := ctx.Args()[0]
 
-	fmt.Printf("Inspeting policy: %s tenant: %s\n", policy, tenant)
+	fmt.Printf("Inspecting policy: %s tenant: %s\n", policy, tenant)
```
Comment (ppadala) [inline netctl/netctl.go]: Have to use the pattern of

log.WithFields(log.Fields{
    "key": "value",
  }).Info

File: netctl/netctl.go [hunk 2]
```diff
@@ -471,7 +471,7 @@ func inspectNetwork(ctx *cli.Context) {
 	tenant := ctx.String("tenant")
 	network := ctx.Args()[0]
 
-	fmt.Printf("Inspeting network: %s tenant: %s\n", network, tenant)
+	fmt.Printf("Inspecting network: %s tenant: %s\n", network, tenant)
```
Comment (ppadala) [inline netctl/netctl.go]: Same here, use log

File: netctl/netctl.go [hunk 3]
```diff
@@ -661,7 +661,7 @@ func inspectEndpointGroup(ctx *cli.Context) {
 	tenant := ctx.String("tenant")
 	endpointGroup := ctx.Args()[0]
 
-	fmt.Printf("Inspeting endpointGroup: %s tenant: %s\n", endpointGroup, tenant)
+	fmt.Printf("Inspecting endpointGroup: %s tenant: %s\n", endpointGroup, tenant)
```
Comment (ppadala) [inline netctl/netctl.go]: Same here, use log

File: netmaster/master/api.go [hunk 4]
```diff
@@ -346,7 +346,7 @@ func UpdateEndpointHandler(w http.ResponseWriter, r *http.Request, vars map[stri
 		return nil, err
 	}
 
-	if epUpdReq.Event == "start" {
+	if epUpdReq.Event == "start" || epUpdReq.Event == "connect" {
```
Comment (ppadala) [inline netmaster/master/api.go]: Can we create const strings (essentially enumeration) instead of naked strings.

File: netplugin/agent/docker_event.go [hunk 5]
```diff
@@ -28,64 +28,80 @@ import (
...
 
 	log.Debugf("Received Docker event: {%#v}\n", *event)
 	endpointUpdReq := &master.UpdateEndpointRequest{}
-	switch event.Status {
-	case "start":
+
+	var containerID string
+	var networkID string
+
+	if (event.Type == "container" && event.Action == "start") ||
```
Comment (ppadala) [inline netplugin/agent/docker_event.go]: same comment here. would be good to get these strings to separate consts.

File: netplugin/agent/docker_event.go [hunk 6]
```diff
@@ -28,64 +28,80 @@ import (
...
+		labelMap := getLabelsFromContainerInspect(&containerInfo)
+		containerTenant := getTenantFromContainerInspect(&containerInfo)
+		networkName, ipAddress, err := getEpNetworkInfoFromContainerInspect(&containerInfo, networkID)
+		if err != nil {
+			log.Errorf("Error getting container network info for %v.Err:%s", containerID, err)
+		}
+		endpoint := getEndpointFromContainerInspect(&containerInfo,
+			networkName, containerTenant)
+
+		if ipAddress != "" {
```
Comment (ppadala) [inline netplugin/agent/docker_event.go]: separate this into another function for ease of reading and maintainability

File: netplugin/agent/docker_event.go [hunk 7]
```diff
@@ -155,14 +183,25 @@ func getContainerFromContainerInspect(containerInfo *types.ContainerJSON) string
...
+	} else {
+		qualifiedName = networkName + "/" + tenantName
+	}
+
 	if containerInfo != nil && containerInfo.NetworkSettings != nil {
-		for _, endpoint := range containerInfo.NetworkSettings.Networks {
+		endpoint, ok := containerInfo.NetworkSettings.Networks[qualifiedName]
+		if ok {
 			endpointID = endpoint.EndpointID
 		}
```
Comment (ppadala) [inline netplugin/agent/docker_event.go]: if !ok, what is in endpoint variable?

File: test/systemtests/aci_util.go [hunk 8]
```diff
@@ -14,7 +14,7 @@ import (
 // EPSpec for aci-gw
 type EPSpec struct {
 	Tenant string `json:"tenant,omitempty"`
-	App    string `json:"app,omitempty"`
+	App    string `json:"app-prof,omitempty"`
```
Comment (ppadala) [inline test/systemtests/aci_util.go]: why the change?

File: test/systemtests/network_test.go [hunk 9]
```diff
@@ -371,6 +435,72 @@ func (s *systemtestSuite) testNetworkAddDeleteTenant(c *C, encap, fwdmode string
...
+				c.Assert(s.cli.NetworkPost(network), IsNil)
+				netNames = append(netNames, network.NetworkName)
+				tenantNames[tenantName] = append(tenantNames[tenantName], network.NetworkName)
+				pktTag++
+			}
+		}
+
+		// Wait 10 seconds for the call backs to come through
+
+		time.Sleep(10 * time.Second)
```
Comment (ppadala) [inline test/systemtests/network_test.go]: this may have timing problems, when we run in different environments. it's better to have a sleep and wait for something to happen in a loop and bail out, after certain number of iterations

File: test/systemtests/util_test.go [hunk 10]
```diff
@@ -1065,6 +1065,35 @@ func (s *systemtestSuite) verifyIPs(ipaddrs []string) error {
...
+/* Confirm network inspect includes a dns endpoint (assumes this is the only endpoint on the network currently) */
+func (s *systemtestSuite) checkNetworkInspectDNS(tenant, network string) (bool, error) {
+
+	netInspect, err := s.cli.NetworkInspect(tenant, network)
+	// Network inspect must succeed
+	if err != nil {
+		return false, err
+	}
+
+	//	logrus.Infof("   %+v   \n", netInspect)
```
Comment (ppadala) [inline test/systemtests/util_test.go]: commented out code, remove or convert it to log.Debug

File: test/systemtests/aci_util.go (hunk 8 above)
Comment (gaurav-dalvi) [inline test/systemtests/aci_util.go]: I needed this small change to work with aci-gw container. I could have opened up another PR. But I asked Sanjeev to incorporate in his PR at that time. This wont cause any problem here.

File: netmaster/master/api.go (hunk 4 above)
Comment (srampal) [inline netmaster/master/api.go]: Just retained the logic/ style being used in the code so far ... but yes, will change it to enumerations.

File: netctl/netctl.go (hunk 2 above)
Comment (srampal) [inline netctl/netctl.go]: same as prior one

File: netctl/netctl.go (hunk 3 above)
Comment (srampal) [inline netctl/netctl.go]: same as prior one

File: netplugin/agent/docker_event.go (hunk 5 above)
Comment (srampal) [inline netplugin/agent/docker_event.go]: same as prior one

Comment (DivyaVavili) [inline netplugin/agent/docker_event.go]: These are docker specific events. Better to have the same constants that docker uses if they are available(like ContainerEventType)

File: test/systemtests/network_test.go (hunk 9 above)
Comment (srampal) [inline test/systemtests/network_test.go]: The same pattern is used throughout the system tests. 10 seconds is an extremely generous time given the responses typically come back in well under 1 second. If we really want to have alternate logic, we will need to look at all the system tests since we have the same pattern in all of them. Hence skipping this suggestion for this change/ fix.

File: netctl/netctl.go (hunk 1 above)
Comment (srampal) [inline netctl/netctl.go]: Per discussion on slack, we shouldn't log here and currently dont log in netctl executable so will just remove the print statement

File: netplugin/agent/docker_event.go (hunk 5 above)
Comment (srampal) [inline netplugin/agent/docker_event.go]: Hi Divya ... sorry, missed replying to this comment earlier. I did look through the docker api (and specially the sam alba code which we use to interface with the docker api). There are no standard docker enums defined and the code just works with json objects (verified with Ranjith as well) so I left the strings as they are currently ... thanks!
